src/legacy/server/logging/rotate/log_rotator.js

```diff
@@ -0,0 +1,316 @@
+
+    if (this.usePolling && this.usePolling !== this.config.get('logging.rotate.usePolling')) {
+      this.log(
+        ['warning', 'logging:rotate'],
+        'The current environment does not support `fs.watch`. Falling back to polling using `fs.watchFile`'
+      );
+    }
+
+    this.stalker = chokidar.watch(this.logFilePath, this._buildWatchCfg(this.usePolling));
+    this.stalker.on('change', async (filename, stats) => await this._logFileSizeMonitorHandler.bind(this)(filename, stats));
+  }
+
+  async _logFileSizeMonitorHandler(filename, stats) {
+    if (!filename || !stats) {
+      return;
+    }
+
+    this.logFileSize = stats.size || 0;
+    await this.throttledRotate();
+  }
+
+  _stopLogFileSizeMonitor() {
+    if (!this.stalker) {
+      return;
+    }
+
+    this.stalker.close();
+    clearTimeout(this.stalkerUsePollingPolicyTestTimeout);
+  }
+
+  _createExitListener() {
+    process.on('exit', this.stop.bind(this));
+  }
+
+  _deleteExitListener() {
+    process.removeListener('exit', this.stop);
+  }
+
+  async _getLogFileSizeAndCreateIfNeeded() {
+    try {
+      const logFileStats = await statAsync(this.logFilePath);
+      return logFileStats.size;
+    } catch {
+      // touch the file to make the watcher being able to register
+      // change events
+      await writeFileAsync(this.logFilePath, '');
+      return 0;
+    }
+  }
+
+  async _callRotateOnStartup() {
+    this.logFileSize = await this._getLogFileSizeAndCreateIfNeeded();
+    await this._rotate();
+  }
+
+  _shouldRotate() {
+    // should rotate evaluation
+    // 1. should rotate if current log size exceeds
+    //    the defined one on everyBytes or has already
+    //    pass the defined time on interval.
+    // 2. should not rotate if is already rotating or if any
+    //    of the conditions on 1. do not apply
+    if (this.isRotating) {
+      return false;
+    }
+
+    return this.logFileSize >= this.everyBytes;
+  }
+
+  async _rotate() {
+    if (!this._shouldRotate()) {
+      return;
+    }
+
+    await this._rotateNow();
+  }
+
+  async _rotateNow() {
+    // rotate process
+    // 1. get rotated files metadata (list of log rotated files present on the log folder, numerical sorted)
+    // 2. delete last file
+    // 3. rename all files to the correct index +1
+    // 4. rename + compress current log into 1
+    // 5. send SIGHUP to reload log config
+
+    // rotate process is starting
+    this.isRotating = true;
+
+    // get rotated files metadata
+    const foundRotatedFiles = await this._readRotatedFilesMetadata();
+
+    // Truncate number of rotated files to the settings limit
+    const rotatedFiles = await this._truncateFoundRotatedFilesToKeepLimit(foundRotatedFiles);
+
+    // delete last file
+    await this._deleteLastRotatedFile(rotatedFiles);
+
+    // rename all files to correct index + 1
+    // and normalize
+    await this._renameRotatedFilesByOne(rotatedFiles);
+
+    // rename + compress current log into 0
```

Comment:
does it do compression?

Reply:
actually it does not, I've decided not to.